// @igabm/n8n-nodes-tiktok/dist/nodes/TikTok/V2/PostStatusDescription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.postStatusFields = exports.postStatusOperations = void 0;
exports.postStatusOperations = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        noDataExpression: true,
        displayOptions: {
            show: {
                resource: ['postStatus'],
            },
        },
        options: [
            {
                name: 'Get',
                value: 'get',
                description: 'Check the status of a post',
                action: 'Get post status',
            },
        ],
        default: 'get',
    },
];
exports.postStatusFields = [
    {
        displayName: 'Publish ID',
        name: 'publishId',
        type: 'string',
        default: '',
        required: true,
        displayOptions: {
            show: {
                resource: ['postStatus'],
                operation: ['get'],
            },
        },
        description: 'Identifier returned when initiating the post upload',
    },
];
