// @igabm/n8n-nodes-tiktok/dist/nodes/TikTok/V2/SearchDescription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.searchFields = exports.searchOperations = void 0;
exports.searchOperations = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        noDataExpression: true,
        required: true,
        displayOptions: {
            show: {
                resource: ['search'],
            },
        },
        options: [
            {
                name: 'Search Hashtags',
                value: 'hashtag',
                description: 'Search for hashtags',
                action: 'Search hashtags',
            },
            {
                name: 'Search Sounds',
                value: 'sound',
                description: 'Search for sounds',
                action: 'Search sounds',
            },
        ],
        default: 'hashtag',
    },
];
exports.searchFields = [
    {
        displayName: 'Query',
        name: 'query',
        type: 'string',
        default: '',
        required: true,
        displayOptions: {
            show: {
                resource: ['search'],
                operation: ['hashtag', 'sound'],
            },
        },
        description: 'Text to search for',
    },
    {
        displayName: 'Cursor',
        name: 'cursor',
        type: 'number',
        default: 0,
        displayOptions: {
            show: {
                resource: ['search'],
                operation: ['hashtag', 'sound'],
            },
        },
        description: 'Cursor for pagination',
    },
    {
        displayName: 'Limit',
        name: 'limit',
        type: 'number',
        typeOptions: { minValue: 1 },
        default: 50,
        displayOptions: {
            show: {
                resource: ['search'],
                operation: ['hashtag', 'sound'],
            },
        },
        description: 'Max number of results to return',
    },
];
